ast/src/parsed/display.rs

```diff
@@ -564,6 +564,90 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
     format!("{}", list.into_iter().format(", "))
 }
 
+impl<E: Display> Expression<E> {
+    pub fn precedence(&self) -> Option<ExpressionPrecedence> {
+        match self {
+            Expression::UnaryOperation(op, _) => Some(op.precedence()),
+            Expression::BinaryOperation(_, op, _) => Some(op.precedence()),
+            _ => None,
+        }
+    }
+
+    pub fn require_parentheses(&self) -> bool {
+        match self {
+            Expression::BinaryOperation(_, op, _) => {
+                op.associativity() == BinaryOperatorAssociativity::RequireParentheses
+            }
+            _ => false,
+        }
+    }
+
+    pub fn format_unary_operation(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Expression::UnaryOperation(op, exp) => {
+                let _exp_string = if let (Some(_precision), Some(_exp_precision)) =
+                    (self.precedence(), exp.precedence())
+                {
+                    if _precision < _exp_precision {
+                        format!("({})", exp)
+                    } else {
+                        format!("{}", exp)
+                    }
+                } else {
+                    format!("{}", exp)
+                };
+
+                if op.is_prefix() {
+                    write!(f, "{}{}", op, _exp_string)
+                } else {
+                    write!(f, "{}{}", _exp_string, op)
+                }
+            }
+            _ => panic!("Format with precedence called on non-unary operation."),
+        }
+    }
+
+    pub fn format_binary_operation(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Expression::BinaryOperation(left, op, right) => {
+                let use_left_parentheses = left.require_parentheses()
+                    || match left.precedence() {
+                        Some(left_precedence) => {
+                            left_precedence > op.precedence()
+                                || (left_precedence == op.precedence()
+                                    && op.associativity() == BinaryOperatorAssociativity::Right)
```

Comment:
Do we really need the `require_parentheses` function? Isn't it just
```suggestion
                let use_left_parentheses = match left.precedence() {
                        Some(left_precedence) => {
                            left_precedence > op.precedence()
                                || (left_precedence == op.precedence()
                                    && op.associativity() != BinaryOperatorAssociativity::Left)
```

Reply:
Generally we won't make anything wrong if we do it in the way I implemented. But in the way you suggested and I think it is like what it was before, I thought there would be a case that the operation precedence be higher than the left expression and the parentheses be skipped. 
So I added that as future proof.

Reply:
Not ethat I changed `== Right` to `!= Left`

Reply:
In our case, since all operations with `RequireParentheses` associativity have the same precedence, my version and your version work the same. But if we had a case which two parentheses required operations with different precedence to be next to each other, there would be a problem.
Overall, I think the current version is more expressive and readable.

Reply:
I may be wrong, but the scenario I imagined is in the future we have two kind of operation with `RequireParentheses` associativity and different precedence. In that case we may not put any parentheses. @chriseth  Do you think this can be probable?

Reply:
It is totally fine to not put parentheses in that case since the precedence is different.